exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/MathFunctionsVarcharInput.java

```diff
@@ -0,0 +1,569 @@
+    }
+
+    public void eval(){
+      String seedStr = org.apache.drill.exec.expr.fn.impl.StringFunctionHelpers.getStringFromVarCharHolder(seed);
+      Double seedDbl = mathUtils.validateInput(seedStr);
+
+      if (seedDbl.isNaN()) {
+        out.value = java.lang.Float.NaN;
+      }
+      else {
+        BigInteger seedBigInt = new java.math.BigInteger(String.valueOf(seedDbl));
+        rand = new java.util.Random(seedBigInt.intValue());
+        out.value = rand.nextDouble();
+      }
+    }
+  }
+  /*
+    // The preexisting power() method accepts VARCHAR input, but this method adds a check for invalid input
+    @FunctionTemplate(name = "power", scope = FunctionTemplate.FunctionScope.SIMPLE, nulls = FunctionTemplate.NullHandling.NULL_IF_NULL)
+    public static class Power implements DrillSimpleFunc{
+
+      @Param VarCharHolder a;
+      @Param VarCharHolder b;
+      @Output  Float8Holder out;
+
+      @Workspace org.apache.drill.exec.expr.fn.impl.MathFunctionsVarcharUtils mathUtils;
+
+      public void setup(){
+        mathUtils = new MathFunctionsVarcharUtils();
+      }
+
+      public void eval(){
+        String aStr = org.apache.drill.exec.expr.fn.impl.StringFunctionHelpers.getStringFromVarCharHolder(a);
+        Double aDbl = mathUtils.validateInput(aStr);
+
+        String bStr = org.apache.drill.exec.expr.fn.impl.StringFunctionHelpers.getStringFromVarCharHolder(b);
+        Double bDbl = mathUtils.validateInput(bStr);
+
+        if (aDbl.isNaN() || bDbl.isNaN()) {
+          out.value = java.lang.Float.NaN;
+        }
+        else {
+          out.value = java.lang.Math.pow(aDbl, bDbl);
+        }
+
+        System.out.println("out.value: " + out.value);
+      }
+
+    }
+
+    // The preexisting power() method accepts VARCHAR input, but this method adds a check for invalid input
+    @FunctionTemplate(name = "power", scope = FunctionTemplate.FunctionScope.SIMPLE, nulls = FunctionTemplate.NullHandling.NULL_IF_NULL)
+    public static class Power implements DrillSimpleFunc{
+
+      @Param VarCharHolder a;
+      @Param Float8Holder b;
+      @Output  Float8Holder out;
+
+      @Workspace org.apache.drill.exec.expr.fn.impl.MathFunctionsVarcharUtils mathUtils;
+
+      public void setup(){
+        mathUtils = new MathFunctionsVarcharUtils();
+      }
+
+      public void eval(){
+        String aStr = org.apache.drill.exec.expr.fn.impl.StringFunctionHelpers.getStringFromVarCharHolder(a);
+        Double aDbl = mathUtils.validateInput(aStr);
+
+        if (aDbl.isNaN()) {
+          out.value = java.lang.Float.NaN;
+        }
+        else {
+          out.value = java.lang.Math.pow(aDbl, b);
+        }
+
+        System.out.println("out.value: " + out.value);
+      }
+
+    }
+
+   */
+  @FunctionTemplate(name = "mod",
+    scope = FunctionTemplate.FunctionScope.SIMPLE,
+    nulls = FunctionTemplate.NullHandling.NULL_IF_NULL)
+  public static class Mod implements DrillSimpleFunc {
+
+    @Param
+    VarCharHolder a;
```

Comment:
These functions are a bit tricky: I wonder if there is a bit of misunderstanding. Drill uses a typed system at execution time. Each math function has declared input types: BIGINT, DOUBLE, VARDECIMAL, etc. The type promotion system will pick the "narrowest" type that is compatible with both arguments, then will cast both arguments, as needed to widen them to the target type. So MOD(BIGINT, BIGINT) is compatible with MOD(B, I) where B is a BIGINT, and I is INT. I is then cast to BIGINT.

As a result, we generate MANY overrides of the same function: one for each of the numeric "families": BIGINT, DOUBLE, VARDECIMAL, with sometimes others thrown in as well, such as maybe an INT version.

So, let's consider this improvement. We want a MOD(VARCHAR, X). What is X? In this bit of code, X is VARCHAR, which matches MOD("10.3", "20"). Cool. But, what about MOD("10.3", 20)? I don't think Drill automatically promotes from numeric types to VARCHAR. (Didn't test, but few SQL systems do.) So, you'd also want MOD(VARCHAR, VARDECIMAL), so the second argument can be cast from the integer and floating point types. To heavy-weight, then add MOD(VARCHAR, BIGINT) and MOD(VARCHAR, DOUBLE). Now the bases are covered.

Except, not yet. The above only handles the second argument. What about the first? So, we also need MOD(BIGINT, VARCHAR), MOD(DOUBLE, VARCHAR), MOD(VARDECIMAL, VARCHAR).

That's a total of 7 implementations for each two-argument math function!

If we did that, I'd almost suggest we break SQL type casting rules and define automatic conversions from VARCHAR to any numeric type. We pick the type based on the auto-cast of VARCHAR (to DOUBLE, say) and any numeric argument. In this case, you get to reuse all the existing math functions for free: no need to change the casting system the hard way.

The result would no longer be SQL compatible. So, maybe a system/session option to enable "Python mode" when needed.

Reply:
@paul-rogers, thank you for your input. I appreciate your comments and thorough explanation. 
We are thinking about these issues and experimenting to see what would work well with Drill and accomplish our goal. 

Please note that this PR is currently marked as `DRAFT` since it is very much a WIP. We can request a review when it is ready.